src/auth/policy/resolver.go

```diff
@@ -1,41 +1,349 @@
+		res = actionEffects.isAuthorized(op)
+		if res.deny != nil {
+			return res
+		}
+	}
+
+	// do we have a prefix matching the action?
+	actionPrefix, _, hasPrefix := rslvr.prefixedActions.LongestPrefix(op.Action)
+	if !hasPrefix {
+		// no prefix so we return
+		return res
+	}
+
+	// we have a prefix
+	// we walk all matching prefix starting from root up to the longest matching prefix
+	// if we meet a single Deny we stop walking and return a Deny
+	rslvr.prefixedActions.WalkPath(actionPrefix, func(_ string, v interface{}) bool {
+		r := v.(*effectResolver).isAuthorized(op)
+		if r.deny != nil {
+			// we met a Deny
+			res = r
+			return true
+		}
+
+		if res.allow == nil {
+			res = r
+		}
+
+		return false
+	})
+
+	return
 }
 
-func NewResolver(policies []types.Policy) (*Resolver, error) {
-	return &Resolver{}, nil
+// effectResolver is associated to each pair path/action (possibly prefixes)
+type effectResolver struct {
+	allow *statement
+	deny  *statement
+}
+
+func (rslvr *effectResolver) isAuthorized(op *Operation) (res *result) {
+	return &result{
+		op:    op,
+		deny:  rslvr.deny,
+		allow: rslvr.allow,
+	}
+}
+
+// addStatement cumulates the effect of multiple statement
+// if a singe statement is Deny, resolver will always resolve to Deny
+// if multiple Deny are passed only the first one is kept
+func (rslvr *effectResolver) addStatement(sttmnt *statement) error {
+	if rslvr.deny != nil {
+		return nil
+	}
+
+	switch sttmnt.Effect {
+	case "Allow":
+		rslvr.allow = sttmnt
+		return nil
+	case "Deny":
+		rslvr.deny = sttmnt
+		rslvr.allow = nil
+		return nil
+	default:
+		return fmt.Errorf("invalid effect %q", sttmnt.Effect)
+	}
+}
+
+// RadixResolver allows to perform authorization checks
+
+// It is built from a set of Policy by aggregating all statements into some
+// radix trees structure allowing optimized authorization checks
+type RadixResolver struct {
+	// exactPath maps an exact path to an action resolver
+	exactPath map[string]*actionsResolver
+
+	// prefixedPath maps prefixed path to an action resolver
+	prefixedPath *radix.Tree
+}
+
+// NewRadixResolver creates a new RadixResolver
+func NewRadixResolver(policies ...*types.Policy) (*RadixResolver, error) {
+	rslvr := &RadixResolver{
+		exactPath:    make(map[string]*actionsResolver),
+		prefixedPath: radix.New(),
+	}
+
+	for _, policy := range policies {
+		err := rslvr.insertPolicy(policy)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return rslvr, nil
+}
+
+// IsAuthorized check whether operation is authorized
+
+// An operation is authorized if
+// - no statement Deny the operation
+// - at least one statement Allow the operation
+func (r *RadixResolver) IsAuthorized(op *Operation) Result {
+	return r.isAuthorized(op)
+}
+
+func (r *RadixResolver) isAuthorized(op *Operation) (res *result) {
+	res = &result{
+		op: op,
+	}
+
+	// do we have an exact match for the path?
+	actionRslvr, hasExact := r.exactPath[op.Path]
```

Comment:
If I would have the following policies, and an user would try to do  `ACTION: keys:Get RESOURCE:stores::stores/mystore/keys/key1`, what would be the result?
```
- effect: Deny
        actions: 
            - keys:Get
        resource:
            - stores::stores/mystore/keys/*
- effect: Allow
        actions: 
            - keys:Get
            - keys:Sign
        resource:
            - stores::stores/mystore/keys/key1
```

Reading from left-to-right I would expect I would expect it to reject, reading right-to-left deny. I personally believe it should be left-to-right, Did you based you decision on how k8s does or any other standard?

Reply:
In the example above it will "Deny"

Reply:
That is right, I would also expect deny due to your first statement, but in the method `isAuthorized(op)` in that particular operation you will fall into the `exactPath` branch of your code:
```
func (r *RadixResolver) isAuthorized(op *Operation) (res *result) {
	res = &result{
		op: op,
	}

	// do we have an exact match for the path?
	actionRslvr, hasExact := r.exactPath[op.Path]
	if hasExact {
		// if we have an exact match and it resolves to Deny we return Deny
		r := actionRslvr.isAuthorized(op)
		if r.deny != nil {
			return r
		}
		res = r
	}
```

what would make it `allowed` if I can read this code propertly